File: client/src/network/protocol.ts
// ── Core type aliases ──────────────────────────────────────────────

export type EntityId = number;
export type Tick = number;

// ── Geometry ───────────────────────────────────────────────────────

export interface Vec2 {
  x: number;
  y: number;
}

// ── Player ─────────────────────────────────────────────────────────

export interface PlayerSnapshot {
  position: Vec2;
  health: number;
  max_health: number;
  tokens: number;
  torch_range: number;
  facing: Vec2;
  dead: boolean;
  death_timer: number;
  attack_cooldown_pct: number;
}

// ── Entities ───────────────────────────────────────────────────────

// Unit-variant enum serializes as plain strings
export type EntityKind = "Agent" | "Building" | "Rogue" | "Item" | "Projectile";

export interface EntityDelta {
  id: EntityId;
  kind: EntityKind;
  position: Vec2;
  data: EntityData;
}

// Serde externally-tagged enum: { "VariantName": { ...fields } }
export type EntityData =
  | { Agent: AgentData }
  | { Building: BuildingData }
  | { Rogue: RogueData }
  | { Item: ItemData }
  | { Projectile: ProjectileData };

export interface AgentData {
  name: string;
  state: AgentStateKind;
  tier: AgentTierKind;
  health_pct: number;
  morale_pct: number;
  stars: number;
  turns_used: number;
  max_turns: number;
  model_lore_name: string;
  xp: number;
  level: number;
}

export interface BuildingData {
  building_type: BuildingTypeKind;
  construction_pct: number;
  health_pct: number;
}

export interface RogueData {
  rogue_type: RogueTypeKind;
  health_pct: number;
}

export interface ItemData {
  item_type: string;
}

export interface ProjectileData {
  dx: number;
  dy: number;
}

// ── Agent enums ────────────────────────────────────────────────────

export type AgentStateKind =
  | "Idle"
  | "Walking"
  | "Building"
  | "Erroring"
  | "Exploring"
  | "Defending"
  | "Critical"
  | "Unresponsive";

export type AgentTierKind =
  | "Apprentice"
  | "Journeyman"
  | "Artisan"
  | "Architect";

// ── Building types ─────────────────────────────────────────────────

export type BuildingTypeKind =
  // Infrastructure
  | "Pylon"
  | "ComputeFarm"
  // Tier 1
  | "TodoApp"
  | "Calculator"
  | "LandingPage"
  // Tier 2
  | "WeatherDashboard"
  | "ChatApp"
  | "KanbanBoard"
  // Tier 3
  | "EcommerceStore"
  | "AiImageGenerator"
  | "ApiDashboard"
  // Tier 4
  | "Blockchain";

// ── Rogue types ────────────────────────────────────────────────────

export type RogueTypeKind =
  | "Corruptor"
  | "Looper"
  | "TokenDrain"
  | "Assassin"
  | "Swarm"
  | "Mimic"
  | "Architect";

// ── Fog of war / chunks ────────────────────────────────────────────

export interface ChunkPos {
  x: number;
  y: number;
}

export interface FogTile {
  light_level: number;
}

// ── Logging ────────────────────────────────────────────────────────

export interface LogEntry {
  tick: Tick;
  text: string;
  category: LogCategory;
}

export type LogCategory =
  | "System"
  | "Agent"
  | "Combat"
  | "Economy"
  | "Exploration"
  | "Building";

// ── Audio ──────────────────────────────────────────────────────────

// Unit-variant enum serializes as plain strings
export type AudioEvent =
  | "AgentSpeak"
  | "CombatHit"
  | "BuildComplete"
  | "RogueSpawn"
  | "CrankTurn"
  | "AgentDeath";

// ── Economy ────────────────────────────────────────────────────────

export interface EconomySnapshot {
  balance: number;
  income_per_sec: number;
  expenditure_per_sec: number;
}

// ── Debug snapshot ─────────────────────────────────────────────────

export interface DebugSnapshot {
  spawning_enabled: boolean;
  god_mode: boolean;
  phase: string;
  crank_tier: string;
}

// ── Project Management ──────────────────────────────────────────────

export interface ProjectManagerState {
  base_dir: string | null;
  initialized: boolean;
  unlocked_buildings: string[];
  building_statuses: Record<string, string>;
  agent_assignments: Record<string, number[]>;
}

// ── Combat events (for client VFX) ────────────────────────────────

export interface CombatEvent {
  x: number;
  y: number;
  damage: number;
  is_kill: boolean;
  rogue_type: RogueTypeKind | null;
}

// ── Main game state update (Server -> Client) ─────────────────────

export interface GameStateUpdate {
  tick: Tick;
  player: PlayerSnapshot;
  entities_changed: EntityDelta[];
  entities_removed: EntityId[];
  fog_updates: [ChunkPos, FogTile[]][];
  economy: EconomySnapshot;
  log_entries: LogEntry[];
  audio_triggers: AudioEvent[];
  debug: DebugSnapshot;
  project_manager: ProjectManagerState | null;
  combat_events: CombatEvent[];
  player_hit: boolean;
  player_hit_damage: number;
}

// ── Server → Client message wrapper ────────────────────────────────
export type ServerMessage =
  | { GameState: GameStateUpdate }
  | { VibeOutput: { agent_id: number; data: number[] } }
  | { VibeSessionStarted: { agent_id: number } }
  | { VibeSessionEnded: { agent_id: number; reason: string } };

// ── Client -> Server messages ──────────────────────────────────────

// Serde externally-tagged enum: unit variants serialize as plain strings,
// struct variants as { "VariantName": { ...fields } }
export type PlayerAction =
  | "Attack"
  | "Interact"
  | "AssignTask"
  | "OpenBuildMenu"
  | { PlaceBuilding: { building_type: BuildingTypeKind; x: number; y: number } }
  | "CrankStart"
  | "CrankStop"
  | "RollbackAgent"
  | { EquipWeapon: { weapon_id: string } }
  | { EquipArmor: { armor_id: string } }
  // Debug actions
  | { DebugSetTokens: { amount: number } }
  | { DebugAddTokens: { amount: number } }
  | "DebugToggleSpawning"
  | "DebugClearRogues"
  | { DebugSetPhase: { phase: string } }
  | { DebugSetCrankTier: { tier: string } }
  | "DebugToggleGodMode"
  | { DebugSpawnRogue: { rogue_type: RogueTypeKind } }
  | "DebugHealPlayer"
  | { DebugSpawnAgent: { tier: AgentTierKind } }
  | "DebugClearAgents"
  // Project management actions
  | { SetProjectDirectory: { path: string } }
  | "InitializeProjects"
  | "ResetProjects"
  | { StartDevServer: { building_id: string } }
  | { StopDevServer: { building_id: string } }
  | { AssignAgentToProject: { agent_id: number; building_id: string } }
  | { UnassignAgentFromProject: { agent_id: number; building_id: string } }
  | "DebugUnlockAllBuildings"
  | "DebugLockAllBuildings"
  | { UnlockBuilding: { building_id: string } }
  | { VibeInput: { agent_id: number; data: string } }
  | { SetMistralApiKey: { key: string } };

export type TaskAssignment =
  | "Build"
  | "Explore"
  | "Guard"
  | "Crank"
  | "Idle";

export interface PlayerInput {
  tick: Tick;
  movement: Vec2;
  action: PlayerAction | null;
  target: EntityId | null;
}

File: server/src/protocol.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

// ── Core type aliases ──────────────────────────────────────────────

pub type EntityId = u64;
pub type Tick = u64;

// ── Geometry ───────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, Serialize, Deserialize, Default)]
pub struct Vec2 {
    pub x: f32,
    pub y: f32,
}

// ── Player ─────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerSnapshot {
    pub position: Vec2,
    pub health: f32,
    pub max_health: f32,
    pub tokens: i64,
    pub torch_range: f32,
    pub facing: Vec2,
}

// ── Entities ───────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum EntityKind {
    Agent,
    Building,
    Rogue,
    Item,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityDelta {
    pub id: EntityId,
    pub kind: EntityKind,
    pub position: Vec2,
    pub data: EntityData,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum EntityData {
    Agent {
        name: String,
        state: AgentStateKind,
        tier: AgentTierKind,
        health_pct: f32,
        morale_pct: f32,
        stars: u8,
        turns_used: u32,
        max_turns: u32,
        model_lore_name: String,
        xp: u64,
        level: u32,
    },
    Building {
        building_type: BuildingTypeKind,
        construction_pct: f32,
        health_pct: f32,
    },
    Rogue {
        rogue_type: RogueTypeKind,
        health_pct: f32,
    },
    Item {
        item_type: String,
    },
}

// ── Agent enums ────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AgentStateKind {
    Idle,
    Walking,
    Building,
    Erroring,
    Exploring,
    Defending,
    Critical,
    Unresponsive,
    Dormant,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum AgentTierKind {
    Apprentice,
    Journeyman,
    Artisan,
    Architect,
}

// ── Building types ─────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum BuildingTypeKind {
    // Infrastructure
    Pylon,
    ComputeFarm,

    // Tier 1
    TodoApp,
    Calculator,
    LandingPage,

    // Tier 2
    WeatherDashboard,
    ChatApp,
    KanbanBoard,

    // Tier 3
    EcommerceStore,
    AiImageGenerator,
    ApiDashboard,

    // Tier 4
    Blockchain,

    // Home Base
    TokenWheel,
    CraftingTable,
}

// ── Rogue types ────────────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum RogueTypeKind {
    Corruptor,
    Looper,
    TokenDrain,
    Assassin,
    Swarm,
    Mimic,
    Architect,
}

// ── Fog of war / chunks ────────────────────────────────────────────

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
pub struct ChunkPos {
    pub x: i32,
    pub y: i32,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize)]
pub struct FogTile {
    pub light_level: f32,
}

// ── Logging ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogEntry {
    pub tick: Tick,
    pub text: String,
    pub category: LogCategory,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum LogCategory {
    System,
    Agent,
    Combat,
    Economy,
    Exploration,
    Building,
}

// ── Audio ──────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum AudioEvent {
    AgentSpeak,
    CombatHit,
    BuildComplete,
    RogueSpawn,
    CrankTurn,
    AgentDeath,
}

// ── Economy ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EconomySnapshot {
    pub balance: i64,
    pub income_per_sec: f64,
    pub expenditure_per_sec: f64,
}

// ── Debug snapshot ─────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DebugSnapshot {
    pub spawning_enabled: bool,
    pub god_mode: bool,
    pub phase: String,
    pub crank_tier: String,
}

// ── Project manager ───────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectManagerState {
    pub base_dir: Option<String>,
    pub initialized: bool,
    pub unlocked_buildings: Vec<String>,
    pub building_statuses: HashMap<String, String>, // building_id -> status string
    pub agent_assignments: HashMap<String, Vec<u64>>, // building_id -> agent entity ids
}

// ── Main game state update (Server → Client) ──────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GameStateUpdate {
    pub tick: Tick,
    pub player: PlayerSnapshot,
    pub entities_changed: Vec<EntityDelta>,
    pub entities_removed: Vec<EntityId>,
    pub fog_updates: Vec<(ChunkPos, Vec<FogTile>)>,
    pub economy: EconomySnapshot,
    pub log_entries: Vec<LogEntry>,
    pub audio_triggers: Vec<AudioEvent>,
    pub debug: DebugSnapshot,
    pub project_manager: Option<ProjectManagerState>,
}

// ── Client → Server messages ───────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum PlayerAction {
    Attack,
    Interact,
    AssignTask,
    OpenBuildMenu,
    PlaceBuilding {
        building_type: BuildingTypeKind,
        x: f32,
        y: f32,
    },
    CrankStart,
    CrankStop,
    RollbackAgent,
    EquipWeapon { weapon_id: String },
    EquipArmor { armor_id: String },

    // Debug actions
    DebugSetTokens { amount: i64 },
    DebugAddTokens { amount: i64 },
    DebugToggleSpawning,
    DebugClearRogues,
    DebugSetPhase { phase: String },
    DebugSetCrankTier { tier: String },
    DebugToggleGodMode,
    DebugSpawnRogue { rogue_type: RogueTypeKind },
    DebugHealPlayer,
    DebugSpawnAgent { tier: AgentTierKind },
    DebugClearAgents,

    // Project management actions
    SetProjectDirectory { path: String },
    InitializeProjects,
    ResetProjects,
    StartDevServer { building_id: String },
    StopDevServer { building_id: String },
    AssignAgentToProject { agent_id: u64, building_id: String },
    UnassignAgentFromProject { agent_id: u64, building_id: String },
    DebugUnlockAllBuildings,
    DebugLockAllBuildings,
    UnlockBuilding { building_id: String },

    // Vibe session actions
    VibeInput { agent_id: u64, data: String },
    SetMistralApiKey { key: String },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub enum TaskAssignment {
    Build,
    Explore,
    Guard,
    Crank,
    Idle,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PlayerInput {
    pub tick: Tick,
    pub movement: Vec2,
    pub action: Option<PlayerAction>,
    pub target: Option<EntityId>,
}

/// Server-to-client message wrapper. All messages sent to the client
/// are wrapped in this enum so the client can distinguish between
/// game state updates and vibe terminal I/O.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum ServerMessage {
    /// Normal game state update (20Hz).
    GameState(GameStateUpdate),
    /// Real-time PTY output from a vibe session.
    VibeOutput { agent_id: u64, data: Vec<u8> },
    /// Vibe session started.
    VibeSessionStarted { agent_id: u64 },
    /// Vibe session ended.
    VibeSessionEnded { agent_id: u64, reason: String },
}
